File: ric_bin/att/grader.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

extern void ricerca(long long int n);

static FILE *file;
static long long int risposta;
static long long int lower_bound = 1;
static long long int upper_bound;
static int subtask;
static long long int min(long long int a, long long int b) { return a < b ? a : b; }
static long long int max(long long int a, long long int b) { return a > b ? a : b; }

int confronto_con(long long int x) {
  if (subtask == 4) {  // se posso cambiare numero segreto (contenuto nella variabile risposta)
    if (x > lower_bound + (upper_bound - lower_bound)/2 )
      risposta = lower_bound;
    else
      risposta = upper_bound;
    fprintf(file, "2 %lld\n", risposta);
  }

  // comunico nel log la "confronto_con" dell'utente
  fprintf(file, "0 %lld\n", x);

  if (x > risposta) {
    upper_bound = min(upper_bound, x - 1);
    return 1;
  } else if (x < risposta) {
    lower_bound = max(lower_bound, x + 1);
    return -1;
  } else {
    return 0;
  }
}

void answer(long long int x) {
  fprintf(file, "1 %lld\n", x);
  exit(0);
}

int main() {

#ifdef EVAL
  file = fopen("input.txt", "r");
#else
  file = stdin;
#endif

  assert( fscanf(file, "%lld %lld %d", &risposta, &upper_bound, &subtask) == 3);
  fclose(file);

#ifdef EVAL
  file = fopen("output.txt", "w");
#else
  file = stdout;
#endif

  ricerca(upper_bound);
  fclose(file);
}

File: ric_bin/sol/grader.py
#!/usr/bin/python3
import sys
import re

subtask = 0
lower_bound = 1
risposta = 0
upper_bound = 0
outfile = None

def confronto_con(x):
    global subtask, outfile, risposta, lower_bound, upper_bound
    print(upper_bound)
    if subtask == 4:  # se posso cambiare numero segreto (contenuto nella variabile risposta)
        if x > lower_bound + (upper_bound - lower_bound) // 2:
            risposta = lower_bound;
        else:
            risposta = upper_bound;
        print("2 %d" % risposta, file = outfile)

    # comunico nel log la "confronto_con" dell'utente
    print("0 %d" % x, file = outfile);

    if x > risposta:
        upper_bound = min(upper_bound, x - 1)
        return 1
    elif x < risposta:
        lower_bound = max(lower_bound, x + 1)
        return -1
    else:
        return 0

def answer(x):
    global outfile
    print("1 %d" % int(x), file = outfile)
    sys.exit(0)

def ottieni_n():
    global subtask, outfile, risposta, upper_bound

    infile = open("input.txt", "r")
    _risposta, _upper_bound, _subtask = [int(s) for s in infile.read().split()]
    risposta = _risposta
    upper_bound = _upper_bound
    subtask = _subtask

    outfile = open("output.txt", "w")
    return upper_bound
